sound/soc/sof/topology.c

```diff
@@ -1256,15 +1256,35 @@ static int sof_connect_dai_widget(struct snd_soc_component *scomp,
 
 		switch (w->id) {
 		case snd_soc_dapm_dai_out:
-			for_each_rtd_cpu_dai(rtd, i, cpu_dai)
-				cpu_dai->capture_widget = w;
+			for_each_rtd_cpu_dai(rtd, i, cpu_dai) {
+				/* find the first empty capture_widget */
+				if (!cpu_dai->capture_widget)
+					break;
```

Comment:
For my understanding: "the first free" means we really don't care which one as long as it's free, or we know, that it will be the right one because they are assigned in the right order and their number matches?

Reply:
> For my understanding: "the first free" means we really don't care which one as long as it's free, or we know, that it will be the right one because they are assigned in the right order and their number matches?

I assume that we will define the dai in topology in the right order. So, suppose that each widget will map to the right BE. However, I think we don't really care if the widget map to the right one since they will be enabled at the same time. 
for example, ideally, SDW1-Pin2 map to ALH0x102.OUT and SDW2-Pin2 map to ALH0x202.OUT. But I think it will also work if SDW1-Pin2 map to ALH0x202.OUT and DW2-Pin2 map to ALH0x102.OUT.

Reply:
is it possible that amp send wrong data channel to speaker ?

Reply:
````
sof-audio-pci  tplg: connected widget ALH2.OUT -> DAI link SDW0-Playback
sof-audio-pci  tplg: connected widget ALH3.IN -> DAI link SDW0-Capture
sof-audio-pci  tplg: connected widget ALH0x102.OUT -> DAI link SDW1-Playback
sof-audio-pci  tplg: connected widget ALH0x302.IN -> DAI link SDW3-Capture
````
so we still can match the widget and dai by name ?

Reply:
@RanderWang Ideally yes we should match. The problem is we already use the stream_name in the widget to match with the BE DAI link name. So what else can we use for matching? Maybe, what we could do is the widget->stream_name should be the same as BE DAI link name which should be the same as the CPU DAI name. Not sure though

Reply:
@ranj063   yes, I got your idea.   I want to match CPU DAI with BE DAI, .e.g.  SDW1 PIN2 <-> ALH0x102.OUT, SDW2 PIN2 <-> ALH0x202.OUT.  Here 0x202 -> CPU DAI 2, PIN2. We can carefully design topology file to keep this principle, but there is a risk.

Reply:
cpu->name is SDWx PINx

Reply:
@RanderWang where will you store the matching information? It will be very convoluted to understand. I think we should just stick to ordering them correctly in tolg

Reply:
> is it possible that amp send wrong data channel to speaker ?

No, we will send identical data to all speakers.

Reply:
@bardliao , your idea is only for 1308, how about other codecs ? if each codec only drives only channel ?

Reply:
> @RanderWang where will you store the matching information? It will be very convoluted to understand. I think we should just stick to ordering them correctly in tolg

here we have cpu dai->name (SDW1 PIN2) and widget->name (ALH0x102.OUT), they should be matched,  and also cpu dai->name (SDW2 PIN2) and widget->name (ALH0x202.OUT),  I don't want a pair like cpu dai->name (SDW1 PIN2) and widget->name (ALH0x202.OUT)

Reply:
> @bardliao , your idea is only for 1308, how about other codecs ? if each codec only drives only channel ?

I believe we will send the same data to all links no matter what codec it is. @plbossart Please correct me if I am wrong.

Reply:
@bard,   you can check https://github.com/thesofproject/linux/blob/topic/sof-dev/Documentation/driver-api/soundwire/stream.rst
````
Example 3: Stereo Stream with L and R channels is rendered by Master. Each of the L and R 

channel is received by two different Slaves. Master and both Slaves are using single port.

+---------------+                    Clock Signal  +---------------+
|    Master     +---------+------------------------+     Slave     |
|   Interface   |         |                        |   Interface   |
|               |         |                        |       1       |
|               |         |           Data Signal  |               |
|    L  +  R    +---+------------------------------+       L       |
|     (Data)    |   |     |    Data Direction      |     (Data)    |
+---------------+   |     |   +------------->      +---------------+
                    |     |
                    |     |
                    |     |                        +---------------+
                    |     +----------------------> |     Slave     |
                    |                              |   Interface   |
                    |                              |       2       |
                    |                              |               |
                    +----------------------------> |       R       |
                                                   |     (Data)    |
                                                   +---------------+
````

Reply:
@bardliao if the Slave device is capable to dealing with multiple channels and re-processing them to generate a different output, we will use this solution as a preferred one.
But for legacy amplifiers designed to deal with a single channel, then we will send only what the Slave driver can deal with.

That was the purpose for my PR #1864, make sure we are ready for all possible topologies and not assume too many things. For example with the Max98063 I think we'll have to use a single channel per amp - to be verified though

Reply:
@RanderWang @plbossart Ok. So we need to ensure DAI widget is connected to the right BE. But we can achieve it by creating DAI widget in the right order. Maybe describe it in the comment is good enough?